File: trabajo 11/merge_sort.js
// Función para fusionar dos subarreglos
function merge(a, l, m, r) {
    const a1 = m - l + 1;
    const a2 = r - m;

    // Crear arreglos temporales
    const L = new Array(a1);
    const R = new Array(a2);

    // Copiar datos a los arreglos temporales
    for (let i = 0; i < a1; i++) {
        L[i] = a[l + i];
    }
    for (let j = 0; j < a2; j++) {
        R[j] = a[m + 1 + j];
    }

    // Fusionar los arreglos temporales de vuelta en a[l..r]
    let x = 0, y = 0, k = l;

    while (x < a1 && y < a2) {
        if (L[x] <= R[y]) {
            a[k] = L[x];
            x++;
        } else {
            a[k] = R[y];
            y++;
        }
        k++;
    }

    // Copiar los elementos restantes de L[], si hay
    while (x < a1) {
        a[k] = L[x];
        x++;
        k++;
    }

    // Copiar los elementos restantes de R[], si hay
    while (y < a2) {
        a[k] = R[y];
        y++;
        k++;
    }
}

// Función principal de Merge Sort
function mergeSort(a, l, r) {
    if (l < r) {
        const m = Math.floor(l + (r - l) / 2);
        mergeSort(a, l, m);
        mergeSort(a, m + 1, r);
        merge(a, l, m, r);
    }
}

// Programa principal
const a = [10, 28, 54, 11];
const s = a.length;

console.log("Arreglo desordenado:");
console.log(a.join(" "));

mergeSort(a, 0, s - 1);

console.log("Arreglo ordenado:");
console.log(a.join(" "));

File: trabajo 11/merge_sort.py
def merge(a, l, m, r):
    a1 = m - l + 1
    a2 = r - m

    # Crear arreglos temporales
    L = [0] * a1
    R = [0] * a2

    # Copiar datos a los arreglos temporales
    for i in range(a1):
        L[i] = a[l + i]
    for j in range(a2):
        R[j] = a[m + 1 + j]

    # Mezclar los arreglos temporales de vuelta en a[l..r]
    i = 0  # Índice para L[]
    j = 0  # Índice para R[]
    k = l  # Índice para a[]

    while i < a1 and j < a2:
        if L[i] <= R[j]:
            a[k] = L[i]
            i += 1
        else:
            a[k] = R[j]
            j += 1
        k += 1

    # Copiar los elementos restantes de L[], si hay
    while i < a1:
        a[k] = L[i]
        i += 1
        k += 1

    # Copiar los elementos restantes de R[], si hay
    while j < a2:
        a[k] = R[j]
        j += 1
        k += 1


def mergesort(a, l, r):
    if l < r:
        m = l + (r - l) // 2  # calcular el punto medio de forma segura
        mergesort(a, l, m)
        mergesort(a, m + 1, r)
        merge(a, l, m, r)


# Programa principal
a = [10, 28, 54, 11]
s = len(a)

print("Arreglo desordenado:")
for j in range(s):
    print("%d" % a[j], end=" ")

# Ordenar con Merge Sort
mergesort(a, 0, s - 1)

print("\nArreglo ordenado:")
for j in range(s):
    print("%d" % a[j], end=" ")
